### PR TITLE
[PLT-7776] Render at-mentions by the teammate name display

### PR DESCRIPTION
**Pull request by philippe-granet**
#### Summary
In the UI, render at-mentions by the teammate name display set in
config.json. The entire name is highlighted, and preceeded with an "@"
symbol.
When writing a message draft, the at-mention still use the username,
similar to channel linking, to guarantee the correct mention is fired.

In system messages (e.g. user joined/left), use the teammate name
display setting as well, instead of the default username.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/7562
https://mattermost.atlassian.net/browse/PLT-7776

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [ ] Added or updated unit tests (required for all new features)
- [ ] Has server changes (please link)
- [ ] Has redux changes (please link)
- [x] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates
- [x] Touches critical sections of the codebase (auth, posting, etc.)

### CODE REVIEW

**Comment by philippe-granet** [comment]
jasonblais Is it normal that test server return a 404 error?

**Comment by jasonblais** [comment]
@philippe-granet not normal, but it seems to be happening for other PRs too. I'll ping one of our engineers to take a look

**Comment by hmhealey** [comment]
I'm going to close and reopen this to trigger a new build. It looks like the web app didn't even get built, so there must've been a hiccup somewhere in the process

**Comment by hmhealey** [comment]
The server works fine now and is ready for testing

**Comment by esethna** [comment]
Thanks @philippe-granet, this looks good! I just wanted to give you a heads up about this PR to add Teammate Name display back to account settings to make sure it won't conflict here:  https://github.com/mattermost/mattermost-webapp/pull/289

**Comment by philippe-granet** [comment]
@esethna I look at PR #289, no conflict, my PR will respect Teammate Name display back to account settings.
I use `Utils.displayUsernameForUser` function which will use user account settings with PR #289 .

**Comment by ccbrown** [comment]
@esethna What kinds of formatting should be applied to these names when they contain hashtags, email addresses, etc.?

At the moment, this PR looks like this:

screen shot 2017-12-08 at 2 49 57 pm

screen shot 2017-12-08 at 2 46 12 pm

So system messages are auto-linking things, while other messages are not. I think I would expect the system messages to just link the full name to the user and not apply any other formatting within the name:

> [Chris sae<</b>b>fd</</b>b>f #asd ~off-topic asd@asd.com](#) has joined the channel.

**Comment by esethna** [comment]
@philippe-granet sorry for the delay in reviewing this PR, we have been focused on release testing. We should have time to do a thorough test pass in the next few days.

**Comment by philippe-granet** [comment]
@jasonblais No news about this PR?

**Comment by esethna** [comment]
Thanks @philippe-granet, overall I think it looks really good, sorry for the delay, there a lot of corner cases to test with this PR. Adding @lindalumitchell to help run a thorough test.

**Comment by lindalumitchell** [comment]
@philippe-granet @esethna I ran some more detailed tests, and looks very good. I am however seeing some system messages that still display the username, regardless of the teammate display setting:

- User joins / leaves a team
- User is removed or added to a channel by another user
- User updates channel header
- User updates channel purpose

The above were tested as part of 3 and 4 in the test outline here (all other scenarios listed in the outline passed without issue): https://mattermost.atlassian.net/browse/PLT-7776?focusedCommentId=43857&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-43857

update-channel-header-or-join-leave-team

added-or-removed-by-other-user

**Comment by esethna** [comment]
@philippe-granet let us know if you have any questions on the above review? ^

**Comment by esethna** [comment]
Thanks @philippe-granet, is this ready to be retested?

**Comment by philippe-granet** [comment]
@esethna Not yet, I only rebase PR branch

**Comment by philippe-granet** [comment]
@esethna This is ready to be retested!

**Comment by philippe-granet** [comment]
@esethna Fix a new bug when nickname, first_name or last_name empty

**Comment by esethna** [comment]
Thanks @philippe-granet! Queuing this for re-testing after we wrap up release testing.

**Comment by lindalumitchell** [comment]
Realizing we have a very old spinmint. Creating a new one.

**Comment by lindalumitchell** [comment]
@philippe-granet @esethna Thank you for your patience on re-testing this one! 

I re-tested (as described in the JIRA ticket), and all the display names are working as expected!

There was just one thing I saw that I'm not sure is an issue (or related to this): It could take a little time, or a channel switch (not sure which affected what I saw) to display user's name after changing the setting in System Console. Here's what I saw:

1. Log in to http://i-00b5b5aaa0d6d9aaf.spinmint.com as system admin (test | passwd, or admin | passwd)
2. Observe the users' names in the NameDisplay team and the PR353 team, Town Square and Off-Topic channels
3. Go to System Console > Users and Teams, and change teammate name display to a different setting than it had been
4. Main Menu > return to a team
5. Observe users' names in Town Square and Off-Topic for that team, making sure to scroll up and Load More Messages to see the beginnings of the channels, with the join messages (should be correct here)
6. Switch to the other team
7. Observe Town Square and Off-Topic, making sure to scroll up and Load More Messages in order to see join messages at the beginning of the channel

Observed: The join messages are missing their users' names in the team you switched to (just says "joined the channel", for example. After some time, or possibly after switching to a different channel and back again, then the users' names appear

Expected: The users' names appear on first load

I do know that this spinmint was built a long time ago now, which is why I had tried to build a new one (but it failed, which is why I could complete testing on this one). Don't know if that could have anything to do with this behavior. Worth getting some help to get a new spinmint going, and I could test it again?

**Comment by lindalumitchell** [comment]
@jasonblais I just realized that Eric is out next week, so maybe you can help take a look at my comment above about a possible issue when you have a chance?

**Comment by jasonblais** [comment]
@lindalumitchell Thanks! That issue should be fixed.

@philippe-granet would you like to help take a look? We really appreciate your efforts here and look forward to merging it soon! 👍

**Comment by lindy65** [comment]
Hi @philippe-granet :)

Just popping in to see how it's going with the updates on this PR?  Let us know if you need any help!

**Comment by philippe-granet** [comment]
@lindalumitchell @jasonblais I can't reproduce this problem. On Spinmint test server, I see this error on console, but I don't see it on my local test server :
```
Uncaught (in promise) TypeError: Cannot read property 'id' of undefined
    at u (channels.js:56)
    at combineReducers.js:120
    at combineReducers.js:120
    at combineReducers.js:120
    at a (configureStore.prod.js:86)
    at t (index.js:21)
    at Array.reduce (<anonymous>)
    at t (index.js:19)
    at updater.js:110
    at autoRehydrate.js:30
```
I rebase my PR from master, could you test again with a new test server?

**Comment by saturninoabril** [comment]
@philippe-granet I recreated the test server and the error is gone.

**Comment by saturninoabril** [comment]
For cases like added/removed user, the display name is still `username` (in the screenshot `jesse.welch` & `adam.torres`).
screen shot 2018-03-01 at 7 33 41 am

I wonder if we could do something on [server](https://github.com/mattermost/mattermost-server/blob/master/app/channel.go#L1115) like:
```
Props: model.StringInterface{
    "userId":      user.Id,
    "addedUserId": addedUser.Id,
},
``` 

Or maybe like below so we'll not break other client/s:
```
Props: model.StringInterface{
    "userId":      user.Id,
    "username":      user.Username,
    "addedUserId": addedUser.Id,
    "addedUsername": addedUser.Username,
},
```

**Comment by jasonblais** [comment]
Thoughts @philippe-granet on the above note?

**Comment by philippe-granet** [comment]
@ccbrown OK, function renderUsernameForUserId added.
@jasonblais I need to do a new PR on mattermost-server project to implement @saturninoabril proposal?

**Comment by jasonblais** [comment]
@saturninoabril Maybe you can help guide @philippe-granet on the next step. We're really close to finishing this, really appreciate your continued help here.

**Comment by philippe-granet** [comment]
@saturninoabril @jasonblais
New PR on mattermost-server https://github.com/mattermost/mattermost-server/pull/8400 to fix added/removed user from Channel/Team
Can you start a new test server with both PR (server and webapp)?

**Comment by lindy65** [comment]
Thanks @philippe-granet :)

@jasonblais and @saturninoabril are both travelling today so should respond to you tomorrow 👍

**Comment by enahum** [comment]
will this be added to https://github.com/mattermost/mattermost-mobile as well?

File: components/post_markdown/system_message_helpers.jsx [hunk 1]
```diff
@@ -46,8 +48,8 @@ function renderLeaveChannelMessage(post, options) {
 }
 
 function renderAddToChannelMessage(post, options) {
-    const username = renderUsername(post.props.username, options);
-    const addedUsername = renderUsername(post.props.addedUsername, options);
+    const username = renderUsernameForUserId(post.user_id, options);
```
Comment (saturninoabril) [inline components/post_markdown/system_message_helpers.jsx]: This will work fine at the moment.
It's not necessary to change on this PR but just like to mention that this will be like below later:
```
const username = renderUsernameForUserId(post.props.userId, options);
```

**Comment by saturninoabril** [comment]
@jasonblais please see @enahum's question above.

@philippe-granet Server PR has been merged recently. Could you check why the user's display name is not rendering correctly? (see screenshot)

screen shot 2018-03-07 at 9 36 06 pm

**Comment by philippe-granet** [comment]
@saturninoabril I think I found the problem with the user's display name not rendering correctly.
When I look at HTTP requests, I see:
- a call to /api/v4/channels/.../posts => used to display current channel posts 
- then a call to /api/v4/users/ids => populate UserStore with all informations about users mentionned in channel

The problem is that UserStore is populated after displaying current channel posts.  
If you go on an other channel and come back on the first channel, you will see that user's display name are rendering correctly, since UserStore has been populated.

A solution could be to add all informations about users (id, name, nickname, firstname, lastname) on server:
```
Props: model.StringInterface{
    "userId":      user.Id,
    "username":      user.Username,
    "nickname":      user.Nickname,
    "firstName":      user.FirstName,
    "lastName":      user.LastName,
....
},
```
Is this a good solution? Is there a better solution for this problem?

**Comment by jasonblais** [comment]
@enahum There's a separate ticket for mobile apps: https://mattermost.atlassian.net/browse/MM-9575

Thanks @philippe-granet for helping with the investigation! Removing `Awaiting Submitter Action` label - @saturninoabril should be around to respond tomorrow

File: components/post_markdown/system_message_helpers.jsx [hunk 2]
```diff
@@ -16,13 +16,15 @@ import PostAddChannelMember from 'components/post_view/post_add_channel_member';
 function renderUsername(value, options) {
     return renderFormattedText(value, {...options, markdown: false});
 }
-
+function renderUsernameForUserId(value, options) {
+    return renderUsername(Utils.displayUsername(value, options));
+}
```
Comment (ccbrown) [inline components/post_markdown/system_message_helpers.jsx]: Can we get some newlines around this function? (I'm surprised this passes our style check.)

**Comment by saturninoabril** [comment]
@philippe-granet I would say no - not to do more on server side - and let's do this on client side. I believe those users are available at the redux store. Maybe need to look at [this](https://github.com/mattermost/mattermost-webapp/blob/master/utils/utils.jsx#L1108) on how we can pull out those ids.

**Comment by philippe-granet** [comment]
@saturninoabril I found this here:
https://github.com/mattermost/mattermost-redux/blob/master/src/actions/posts.js#L575

Users profiles are populated asynchronously in redux store. They are not yet available on first channel display.
```js
export function getPosts(channelId, page = 0, perPage = Posts.POST_CHUNK_SIZE) {
    return async (dispatch, getState) => {
        dispatch({type: PostTypes.GET_POSTS_REQUEST}, getState);
        let posts;

        try {
            posts = await Client4.getPosts(channelId, page, perPage);
            getProfilesAndStatusesForPosts(posts.posts, dispatch, getState);
...
```

**Comment by saturninoabril** [comment]
@philippe-granet Thanks! I'm good with the fallback. 
@jasonblais I'm all good with the changes and will merge this PR.  Note that we have [MM-9780](https://mattermost.atlassian.net/browse/MM-9780) to take a look further on the issue of missing teammate name display setting.

**Comment by jasonblais** [comment]
@philippe-granet Thanks so much for your help here!

**Comment by jasonblais** [comment]
@philippe-granet I was curious if you were interested adding support for this in the mobile apps too? https://mattermost.atlassian.net/browse/MM-9575
